### PR TITLE
Add the ip6tables service

### LINKED ISSUE #6351: zen does not setup ipv6 firewall
**Describe the bug**

While doing a network scan, I noticed that packetfence did answered some services through IPv6.
I checked and ip6tables is clean:

```
[root@vms188l etc]# ip6tables -L
Chain INPUT (policy ACCEPT)
target     prot opt source               destination         

Chain FORWARD (policy ACCEPT)
target     prot opt source               destination         

Chain OUTPUT (policy ACCEPT)
target     prot opt source               destination   
```      

Shouldn't packetfence be responsible to, at least, block incoming IPv6 traffic?

I'm using zen running with packetfence-10.2.0-20201007132254.199263253.0007.v10.2.0.el7.x86_64.
It is not the latest but I didn't find anything related in changelog.

**To Reproduce**
* Setup IPv6 address using packetfence web interface.
* check ip6tables

### PR DESCRIPTION
**Pull request by julsemaan**
# Description
Adds the ip6tables service

# Impacts
New service

# Issue
fixes #6351 

# Delete branch after merge
YES

# Checklist
(REQUIRED) - [yes, no or n/a]
- [N/A] Document the feature
- [N/A] Add unit tests
- [TODO] Add acceptance tests (TestLink)

# NEWS file entries
## Enhancements
* The ip6tables rules are now managed by PacketFence

### CODE REVIEW

**Comment by julsemaan** [comment]
Tested the packaging on Debian + RHEL and its fine

We'll have an acceptance test in the integration tests with a little magic from @nqb

**Review by nqb — CHANGES_REQUESTED** [review]
Packaging looks good. 

Regarding binary location for Debian, I think we just need to apply a patch on `pf.conf` like we do for FreeRADIUS: https://github.com/inverse-inc/packetfence/blob/84892e085adc6b342790f805bc61fc9361f7a149/debian/patches/debianize.patch#L42-L49

to have a working ip6tables on Debian without any visible override on `pf.conf`

File: conf/documentation.conf [hunk 1]
```diff
@@ -465,6 +472,13 @@ Location of the iptables binary. Only necessary to change if you are
 not running the RPMed version.
 EOT
 
+[services.ip6tables_binary]
+type=hidden
+description=<<EOT
+Location of the ip6tables binary. Only necessary to change if you are
+not running the RPMed version.
```
Comment (extrafu) [inline conf/documentation.conf]: "RPMed version" - how about the Debian packages? Should we simply say "if you are not running the official packaged version" ?

File: conf/pf.conf.defaults [hunk 2]
```diff
@@ -451,6 +451,16 @@ iptables=enabled
...
 iptables_binary=/sbin/iptables
 #
+# services.ip6tables
+#
+# Should ip6tables be managed by PacketFence?
+ip6tables=enabled
+# 
+# services.ip6tables_binary
+# 
+# location of the ip6tables binary. only necessary to change if you are not running the rpmed version.
```
Comment (extrafu) [inline conf/pf.conf.defaults]: Same as above

**Comment by julsemaan** [comment]
> Packaging looks good.
> 
> Regarding binary location for Debian, I think we just need to apply a patch on `pf.conf` like we do for FreeRADIUS:
> 
> https://github.com/inverse-inc/packetfence/blob/84892e085adc6b342790f805bc61fc9361f7a149/debian/patches/debianize.patch#L42-L49
> 
> to have a working ip6tables on Debian without any visible override on `pf.conf`

Based on my findings, the binary is located at the same place on RHEL 8 and Debian 11

File: lib/pf/ip6tables.pm [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+);
+use pf::file_paths qw($generated_conf_dir $conf_dir);
+
+# This is the content that needs to match in the iptable rules for the service
+# to be considered as running
+Readonly our $FW_FILTER_INPUT_MGMT      => 'input-v6-management-if';
+Readonly our $FW_FILTER_INPUT_PORTAL    => 'input-v6-portal-if';
+
+sub generate {
+    my ($class) = @_;
```
Comment (fdurand) [inline lib/pf/ip6tables.pm:40]: self instead of class

File: lib/pf/ip6tables.pm [hunk 4]
```diff
@@ -0,0 +1,109 @@
...
+EOF
+    }
+
+    parse_template( \%tags, "$conf_dir/ip6tables.conf", "$generated_conf_dir/ip6tables.conf" );
+    $class->restore("$generated_conf_dir/ip6tables.conf");
+}
+
+
+sub save {
+    my ($class, $save_file) = @_;
```
Comment (fdurand) [inline lib/pf/ip6tables.pm:63]: self instead of class

File: lib/pf/ip6tables.pm [hunk 5]
```diff
@@ -0,0 +1,109 @@
...
+    my $logger = get_logger();
+    $logger->info( "saving existing ip6tables to " . $save_file );
+    pf_run("/usr/sbin/ip6tables-save -t nat > $save_file");
+    pf_run("/usr/sbin/ip6tables-save -t mangle >> $save_file");
+    pf_run("/usr/sbin/ip6tables-save -t filter >> $save_file");
+}
+
+sub restore {
+    my ($class, $restore_file) = @_;
```
Comment (fdurand) [inline lib/pf/ip6tables.pm:72]: self instead of class

File: lib/pf/services/manager/ip6tables.pm [hunk 6]
```diff
@@ -0,0 +1,166 @@
...
+
+=cut
+
+sub startService {
+    my ($self) = @_;
+    unless ($self->isAlive()) {
+        pf::ip6tables->save($install_dir . '/var/ip6tables.bak');
+    }
+    pf::ip6tables->generate();
+    return 1;
```
Comment (fdurand) [inline lib/pf/services/manager/ip6tables.pm]: Use constant

File: lib/pf/services/manager/ip6tables.pm [hunk 7]
```diff
@@ -0,0 +1,166 @@
...
+
+=head2
+
+generateConfig
+
+=cut
+
+sub generateConfig {
+    pf::ip6tables->generate();
+    return 1;
```
Comment (fdurand) [inline lib/pf/services/manager/ip6tables.pm]: Use constant

File: lib/pf/services/manager/ip6tables.pm [hunk 8]
```diff
@@ -0,0 +1,166 @@
...
+=head2 stop
+
+Wrapper around systemctl. systemctl should in turn call the actual _stop.
+
+=cut
+
+sub stop {
+    my ($self) = @_;
+    system('sudo systemctl stop packetfence-ip6tables');
+    return 1;
```
Comment (fdurand) [inline lib/pf/services/manager/ip6tables.pm]: Use constant

File: lib/pf/services/manager/ip6tables.pm [hunk 9]
```diff
@@ -0,0 +1,166 @@
...
+
+stop ip6tables (called from systemd)
+
+=cut
+
+sub _stop {
+    my ($self) = @_;
+    my $logger = get_logger();
+    pf::ip6tables->restore( $install_dir . '/var/ip6tables.bak' );
+    return 1;
```
Comment (fdurand) [inline lib/pf/services/manager/ip6tables.pm]: Use constant

File: lib/pf/services/manager/ip6tables.pm [hunk 10]
```diff
@@ -0,0 +1,166 @@
...
+=cut
+
+sub isAlive {
+    my ($self) = @_;
+    my $logger = get_logger();
+    my $result;
+    my $pid = $self->pid;
+    my $_EXIT_CODE_EXISTS = "0";
+    my $rules_applied = defined( pf_run( "sudo " . $Config{'services'}{"ip6tables_binary"} . " -S | grep " . $pf::ip6tables::FW_FILTER_INPUT_MGMT ,accepted_exit_status => [$_EXIT_CODE_EXISTS]) );
+    return ($pid && $rules_applied) ? 1 : 0;
```
Comment (fdurand) [inline lib/pf/services/manager/ip6tables.pm]: Use constant

File: lib/pf/services/manager/ip6tables.pm [hunk 11]
```diff
@@ -0,0 +1,166 @@
...
+=head2 start
+
+Wrapper around systemctl. systemctl should in turn call the actuall _start.
+
+=cut
+
+sub start {
+    my ($self,$quick) = @_;
+    system('sudo systemctl start packetfence-ip6tables');
+    return $? == 0;
```
Comment (fdurand) [inline lib/pf/services/manager/ip6tables.pm]: Use constant

**Review by fdurand — CHANGES_REQUESTED** [review]
It miss the debianize.patch change.
Do we need to log in ip6tables ?

File: lib/pf/ip6tables.pm (hunk 3 above)
Comment (julsemaan) [reply lib/pf/ip6tables.pm:40]: AFAIK, we use `$class` when we don't have to call `new` on the module (its the case here)

@jrouzierinverse can correct me if I'm wrong

**Comment by julsemaan** [comment]
> It miss the debianize.patch change. 

AFAIK there is nothing necessary there because the paths are the same on Debian and RHEL (I tested a package on both)
That is unless something else than the binary paths has to be debianized

> Do we need to log in ip6tables ?

You mean ip6tables.pm? Or the actual ip6tables chain?

**Comment by fdurand** [comment]
> > It miss the debianize.patch change.
> 
> AFAIK there is nothing necessary there because the paths are the same on Debian and RHEL (I tested a package on both) That is unless something else than the binary paths has to be debianized

It´s for the systemd script, the perl path is not the same.
> 
> > Do we need to log in ip6tables ?
> 
> You mean ip6tables.pm? Or the actual ip6tables chain?

I mean that:
-j LOG --log-prefix "STATE_INVALID_DROP: "

**Comment by julsemaan** [comment]
I've moved the venom tests into [fix/6351](https://github.com/inverse-inc/packetfence/tree/fix/6351-test)

**Comment by julsemaan** [comment]
> > > It miss the debianize.patch change.
> > 
> > 
> > AFAIK there is nothing necessary there because the paths are the same on Debian and RHEL (I tested a package on both) That is unless something else than the binary paths has to be debianized
> 
> It´s for the systemd script, the perl path is not the same.

According to my Debian 11 machine, `/bin/perl` points to the Perl binary:
```
root@test-pf:~# /bin/perl --version

This is perl 5, version 32, subversion 1 (v5.32.1) built for x86_64-linux-gnu-thread-multi
(with 46 registered patches, see perl -V for more detail)

Copyright 1987-2021, Larry Wall

Perl may be copied only under the terms of either the Artistic License or the
GNU General Public License, which may be found in the Perl 5 source kit.

Complete documentation for Perl, including FAQ lists, should be found on
this system using "man perl" or "perldoc perl".  If you have access to the
Internet, point your browser at http://www.perl.org/, the Perl Home Page.
```

With that in mind, is it really worth changing debianize.patch for patching something that isn't required

> > > Do we need to log in ip6tables ?
> > 
> > 
> > You mean ip6tables.pm? Or the actual ip6tables chain?
> 
> I mean that: -j LOG --log-prefix "STATE_INVALID_DROP: "

I'll remove this

**Comment by fdurand** [comment]
If the perl path is ok on debian 11 then we are good so.

**Comment by julsemaan** [comment]
Everything should be adjusted except the `$self` vs `$class` which is due to calling the functions on the class and not on a blessed instance. 

Happy to change it if we state we must use `$self` even in this case

Some examples of `$class`:
- lib/pf/api.pm
- lib/pf/Connection/Profile.pm
- lib/pf/Portal/Session.pm

**Pull request merged**